packages/ketcher-react/src/script/editor/operations/text.ts

```diff
@@ -0,0 +1,213 @@
+/****************************************************************************
+ * Copyright 2021 EPAM Systems
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
+import { BaseOperation } from './base'
+import { OperationType } from './OperationType'
+import { Text, Vec2, scale } from 'ketcher-core'
+import Restruct, { ReText } from '../../render/restruct'
+
+interface TextCreateData {
+  id?: any
+  label: string
+  position: Vec2
+  type: string
+}
+
+export class TextCreate extends BaseOperation {
+  data: TextCreateData
+  performed: boolean
+
+  constructor(id: any, label: string = '', position: Vec2, type: string) {
+    super(OperationType.TEXT_CREATE)
+    this.data = { id, label, position, type }
+    this.performed = false
+  }
+
+  execute(restruct: Restruct): void {
+    const struct = restruct.molecule
+
+    if (!this.performed) {
+      this.data.id = struct.texts.add(new Text(this.data))
+      this.performed = true
+    } else {
+      struct.texts.set(this.data.id, new Text(this.data))
+    }
+
+    restruct.texts.set(this.data.id, new ReText(struct.texts.get(this.data.id)))
+
+    const { id, position } = this.data
+
+    struct.textSetPosition(id, new Vec2(position))
+
+    BaseOperation.invalidateItem(
+      restruct,
+      'texts',
+      // @ts-ignore
+      this.data.id,
+      1
+    )
+  }
+
+  invert(): BaseOperation {
+    //@ts-ignore
+    return new TextDelete(this.data.id)
+  }
+}
+
+interface TextUpdateData {
+  id?: any
+  label?: string
+  position?: Vec2
+  type?: string
+}
+
+export class TextUpdate extends BaseOperation {
+  oldData: TextUpdateData
+  newData: TextUpdateData | null
+
+  constructor(id?: any, label?: string, position?: Vec2, type?: string) {
+    super(OperationType.TEXT_UPDATE)
+    this.oldData = { id, label, position, type }
```

Comment:
Please delete oldData and newData according our agreements

Reply:
I've refactored this part a bit & tested multiple steps inversion (up to 30), it is working correctly. 

I would appreciate it if you give me a piece of advice on how to get rid off newData/oldData pair, as I was not able to do so.

Reply:
refactored